src/uct/ib/rc/accel/rc_mlx5_devx.c

```diff
@@ -0,0 +1,196 @@
+    uct_ib_mlx5_srq_seg_t *seg;
+    struct mlx5dv_pd dvpd = {};
+    struct mlx5dv_cq dvcq = {};
+    struct mlx5dv_obj dv = {};
+    void *xrqc, *wq;
+    int len, ret, max, i;
+
+    uct_rc_mlx5_init_rx_tm_common(iface, sizeof(struct ibv_rvh));
+
+    max = ucs_max(config->super.rx.queue_len, IBV_DEVICE_MIN_UWQ_POST);
+    max = ucs_roundup_pow2(max);
+    len = max * UCT_IB_MLX5_SRQ_STRIDE;
+    ret = posix_memalign(&iface->rx.srq.buf, ucs_get_page_size(), len);
+    if (ret) {
+        return status;
+    }
+
+    iface->rx.srq.devx.mem = mlx5dv_devx_umem_reg(dev->ibv_context,
+                                                  iface->rx.srq.buf, len,
+                                                  IBV_ACCESS_LOCAL_WRITE);
+    if (!iface->rx.srq.devx.mem) {
+        goto err_free_buf;
+    }
+
+    iface->rx.srq.devx.dbrec = ucs_mpool_get_inline(&md->dbrec_pool);
+    if (!iface->rx.srq.devx.dbrec) {
+        goto err_free_mem;
+    }
+
+    iface->rx.srq.db = &iface->rx.srq.devx.dbrec->db[MLX5_RCV_DBR];
+    dv.pd.in         = uct_ib_iface_md(&iface->super.super)->pd;
+    dv.cq.in         = iface->super.super.cq[UCT_IB_DIR_RX];
+    dv.pd.out        = &dvpd;
+    dv.cq.out        = &dvcq;
+    mlx5dv_init_obj(&dv, MLX5DV_OBJ_PD | MLX5DV_OBJ_CQ);
+
+    UCT_IB_MLX5DV_SET(create_xrq_in, in, opcode, UCT_IB_MLX5_CMD_OP_CREATE_XRQ);
+    xrqc = UCT_IB_MLX5DV_ADDR_OF(create_xrq_in, in, xrq_context);
+
+    UCT_IB_MLX5DV_SET(xrqc, xrqc, topology, UCT_IB_MLX5_XRQC_TOPOLOGY_TAG_MATCHING);
+    UCT_IB_MLX5DV_SET(xrqc, xrqc, offload,  UCT_IB_MLX5_XRQC_OFFLOAD_RNDV);
+    UCT_IB_MLX5DV_SET(xrqc, xrqc, tag_matching_topology_context.log_matching_list_sz,
+                                  ucs_ilog2(iface->tm.num_tags));
+    UCT_IB_MLX5DV_SET(xrqc, xrqc, dc,       1);
+    UCT_IB_MLX5DV_SET(xrqc, xrqc, cqn,      dvcq.cqn);
+
+    wq = UCT_IB_MLX5DV_ADDR_OF(xrqc, xrqc, wq);
+    UCT_IB_MLX5DV_SET(wq, wq, log_wq_sz,     ucs_ilog2(max));
+    UCT_IB_MLX5DV_SET(wq, wq, log_wq_stride, ucs_ilog2(UCT_IB_MLX5_SRQ_STRIDE));
+    UCT_IB_MLX5DV_SET(wq, wq, pd,            dvpd.pdn);
+    UCT_IB_MLX5DV_SET(wq, wq, dbr_umem_id,   iface->rx.srq.devx.dbrec->mem_id);
+    UCT_IB_MLX5DV_SET64(wq, wq, dbr_addr,    iface->rx.srq.devx.dbrec->offset);
+    UCT_IB_MLX5DV_SET(wq, wq, wq_umem_id,    iface->rx.srq.devx.mem->umem_id);
+
+    iface->rx.srq.devx.obj = mlx5dv_devx_obj_create(dev->ibv_context, in, sizeof(in),
+                                                    out, sizeof(out));
+    if (iface->rx.srq.devx.obj == NULL) {
+        ucs_error("mlx5dv_devx_obj_create(SRQ) failed, syndrome %x: %m",
+                  UCT_IB_MLX5DV_GET(create_xrq_out, out, syndrome));
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    iface->rx.srq.type    = UCT_IB_MLX5_OBJ_TYPE_DEVX;
+    iface->rx.srq.srq_num = UCT_IB_MLX5DV_GET(create_xrq_out, out, xrqn);
+    iface->tm.cmd_qp_len  = (2 * iface->tm.num_tags);
+
+    iface->rx.srq.free_idx      = max - 1;
```

Comment:
looks like the SRQ initialization code below should be common with non-devx (e.g mlx5dv/mlx5_hw.h) without devx?

Reply:
unhandled comment